### ISSUE TITLE
NiftiMasker.transform() "eats" big part of the target image data

### ISSUE THREAD

**Issue by banilo** [nilearn/nilearn#858]
Just doing a simple a masker.transform() on this resting-state image (high values in the xE7 area) actually crushed most image values down to zero in the output (transformed) matrix. Appears to be a type-casting issue, perhaps in the NiftiMasker class? Any ideas?

```python
 masker.inverse_transform(masker.transform('rsfix_img732.nii')).to_filename('rsfix_img732_test.nii')
```

Here are the input files:
https://dl.dropboxusercontent.com/u/4403154/MNI152GM.nii
https://dl.dropboxusercontent.com/u/4403154/rsfix_img732_test.nii
https://dl.dropboxusercontent.com/u/4403154/rsfix_img732.nii

<img width="827" alt="masker_pacman" src="https://cloud.githubusercontent.com/assets/3907396/11400972/a3131f4c-9390-11e5-8467-3cf2755830b1.png">

**Comment by banilo** [comment]
Addendum: If I apply the mask by hand, all values remain in the output nifti. So there is some quirky thresholding/type-casting going on in the NiftMasker class.

```python
data = nib.load(rs_files[ind]).get_data()[masker_bool]
masker.inverse_transform(data).to_filename('rsfix_img732_test_bool.nii')
```

**Comment by GaelVaroquaux** [comment]
Could you post the full set of lines that enables reproducing? I need the
lines with which you instanciate the masker (to know the options that you
are using), and how you fit it.

**Comment by bthirion** [comment]
The problems comes from the fact that the affine of the RS image is wrong:
In [36]: nib.load('rsfix_img732.nii').get_affine()
Out[36]: 
array([[ 1.,  0.,  0.,  1.],
       [ 0.,  1.,  0.,  1.],
       [ 0.,  0.,  1.,  1.],
       [ 0.,  0.,  0.,  1.]])

In [37]: masker.mask_img_.get_affine()
Out[37]: 
array([[  -2.,    0.,    0.,   90.],
       [   0.,    2.,    0., -126.],
       [   0.,    0.,    2.,  -72.],
       [   0.,    0.,    0.,    1.]])

**Comment by banilo** [comment]
@GaelVaroquaux: The NiftiMasker instantiation is completely standard - just the mask image is specified in the constructor, then fit().

**Comment by banilo** [comment]
Good to know that the affine is probably the culprit. Some users might expect the masking/transforming to take place in voxel space (and assume affine does not play a role), rather than brain space (and assume affine is relevant).

**Comment by GaelVaroquaux** [comment]
> @GaelVaroquaux: The NiftiMasker instantiation is completely standard -
> just the mask image is specified in the constructor, then fit().

Please do give me the lines. My hunch is that you are making a user error
here. So I do need the lines to actually check.

**Comment by banilo** [comment]
Here we go with the code:

```python
masker = NiftiMasker(mask_img='MNI152GM.nii')
masker.fit()
```

**Comment by AlexandreAbraham** [comment]
@GaelVaroquaux the culprit is the affine, no need to investigate further.

**Comment by GaelVaroquaux** [comment]
> masker = NiftiMasker(mask_img='MNI152GM.nii')
> masker.fit()

OK. Then I don't understand why you are complaining: your mask image and
your data only partly overlap. The NiftiMasker is doing what it is
supposed to do.

**Comment by GaelVaroquaux** [comment]
I closed the issue, but I worry a bit about user support.

We are always going to have users who feed in erroneous masks, and then complain that it does work.

If even @banilo, who is an advanced user (a developer even), opens an issue on the tracker when he does this, I fear that many more will.

A very simply plotting "plotting.plot_stat_map('rsfix_img732.nii', 'MNI152GM.nii')" can diagnose the problem immediately, but people don't do it. That goes back to the idea that we need reporting or diagnosing utilities, with such plots.

**Comment by banilo** [comment]
Well, as proposed above, you might simply state explicitly in
docs/docstring that the masking (the "overlap") is examined in brain space
and that the affines need to be healthy for that.

Up to now, I though it is a simple "binary" operation in voxel space -
matrix matching. That is by the way what probably most people in matlab do,
who do not have a dedicated API such as nilearn does.


2015-11-27 7:38 GMT+01:00 Gael Varoquaux <notifications@github.com>:

**Comment by GaelVaroquaux** [comment]
> Well, as proposed above, you might simply state explicitly in
> docs/docstring that the masking (the "overlap") is examined in brain space
> and that the affines need to be healthy for that.

Sure. PR welcomed.

> Up to now, I though it is a simple "binary" operation in voxel space -
> matrix matching.

Every single operation in nilearn is implemented in brain space.

**Comment by banilo** [comment]
> Every single operation in nilearn is implemented in brain space.

This appears to be an overarching design principle that is worth mentioning somewhere at the very beginning at nilearn. How a about a short list à la "Overarching design principles":
* nifti image is the primary working unit, not 3D/4D matrices
* any comparisons/manipulation is done in world space, not voxel space
* transform() performs a manipulation, inverse_transform() reverses that manipulation
* access to standard datasets (e.g. region/network atlases) is pipelined by fetchers
* ...
* ...

**Comment by GaelVaroquaux** [comment]
> This appears to be an overarching design principle that is worth mentioning
> somewhere at the very beginning at nilearn.

Sure. Where would you put this?

**Comment by banilo** [comment]
> Sure. Where would you put this?

Somewhere at the very beginning of the docs.

Or even more drastically think about ~5 core design principles of nilearn usage (such the onces mentioned above) that are printed to the console by default after ``import nilearn``...

**Comment by GaelVaroquaux** [comment]
On Thu, Nov 26, 2015 at 11:07:43PM -0800, Danilo Bzdok wrote:
>     Sure. Where would you put this?

> Somewhere at the very beginning of the docs.

I am OK with a sidebar on this specific aspect in the getting started,
linking the page in "manipulating images" that explains what we mean by a
niimg.
Not anything more: it would be cluttering.

> Or even more drastically think about ~5 core design principles of
> nilearn usage (such the onces mentioned above) that are printed to the
> console by default after import nilearn...

I'd rather not.

**Comment by eickenberg** [comment]
hmm what is the point of equipping images with affines if they are then not
used?

maybe in this concrete case we can actually issue a warning: if the mask
shape and signal shape are exactly equal but affines differ, then warn
saying that nilearn considers these images different but that the user may
think they are the same. (Or if the mask is passed as just an array, go
voxelwise ..... I know, scary)

Alternatively, using the identity matrix as affine is improbable in real
life. When it happens, one could also directly warn (although note that I
would get quite angry at nilearn incessantly commenting on my choice of
affine, if it was deliberate)

**Comment by AlexandreAbraham** [comment]
A dedicated page "from matlab to nilearn" would be the perfect place for that kind of information.

Also, I guess that `_check_same_fov` is called in the transform. We could raise a specific warning when the shape is the same but the affine is different, especially if the shape is dummy.

**Comment by GaelVaroquaux** [comment]
> maybe in this concrete case we can actually issue a warning: if the mask
> shape and signal shape are exactly equal but affines differ, 

> Alternatively, using the identity matrix as affine is improbable in real
> life. 

I think that doing an 'and' on these two conditions, and a warning would
be a good thing. +1 for a PR on that.

**Comment by GaelVaroquaux** [comment]
> A dedicated page "from matlab to nilearn" would be the perfect place
> for that kind of information.

I think here it would be a 'from FSL to nilearn' rather. But indeed, such
a page would be great.

> Also, I guess that _check_same_fov is called in the transform. We could
> raise a specific warning when the shape is the same but the affine is
> different, especially if the shape is dummy.

That's the right place to implement the warning of @eickenberg

**Comment by meghalD** [comment]
I tried Abide Dataset and ran the following code to fit and transform NiftiMasker.

_masker = NiftiMasker(mask_img=mask_filename, smoothing_fwhm=4,
                     standardize=True, memory="nilearn_cache", memory_level=1)
func_filename = haxby_dataset.func[0]
X = masker.fit_transform(func_filename)

##error : Input data has incompatible dimensionality, Expected 3D image and you provided a list of 3D images(4D)

**Comment by bthirion** [comment]
We're missing a bit of context here. what is  `mask_filename` here ?

**Comment by meghalD** [comment]
> We're missing a bit of context here. what is `mask_filename` here ?

'mask_filename' as the name suggest is the mask file obtained while downloading the dataset. In my case I obtained it as follows - 

abide_dataset = datasets.fetch_abide_pcp(derivatives=['func_mean','func_preproc','func_mask'],
n_subjects=10)
mask_filenames = abide_dataset.func_mask

**Comment by bthirion** [comment]
OK. Please let me know which example you're running exactly.

**Comment by meghalD** [comment]
> OK. Please let me know which example you're running exactly.

http://nilearn.github.io/auto_examples/02_decoding/plot_haxby_anova_svm.html

This is the tutorial I was referring to.

**Comment by meghalD** [comment]
> This is a minimal example to show how to mask ABIDE data using `NiftiMasker`. Note that I used single subject.
> 
> May be this helps you get started.
> 
> ```python
> from nilearn import datasets
> abide = datasets.fetch_abide_pcp(derivatives=['func_mean','func_preproc','func_mask'], 
>                                  n_subjects=1)
> mask_filename = abide.func_mask[0]
> func_filename = abide.func_preproc[0]
> 
> from nilearn.input_data import NiftiMasker
> masker = NiftiMasker(mask_img=mask_filename, smoothing_fwhm=4,
>                      standardize=True, memory="nilearn_cache", memory_level=1)
> X = masker.fit_transform(func_filename)
> ```

Thank you ! This worked for me.

**Comment by meghalD** [comment]
> This is a minimal example to show how to mask ABIDE data using `NiftiMasker`. Note that I used single subject.
> 
> May be this helps you get started.
> 
> ```python
> from nilearn import datasets
> abide = datasets.fetch_abide_pcp(derivatives=['func_mean','func_preproc','func_mask'], 
>                                  n_subjects=1)
> mask_filename = abide.func_mask[0]
> func_filename = abide.func_preproc[0]
> 
> from nilearn.input_data import NiftiMasker
> masker = NiftiMasker(mask_img=mask_filename, smoothing_fwhm=4,
>                      standardize=True, memory="nilearn_cache", memory_level=1)
> X = masker.fit_transform(func_filename)
> ```

Can you tell me how to build condition mask for Abide dataset; as there is no "behaviour" field in the dataset.

import pandas as pd

# Load target information as string and give a numerical identifier to each
behavioral = pd.read_csv(haxby_dataset.session_target[0], sep=" ")
conditions = behavioral['labels']

# Restrict the analysis to faces and places
condition_mask = behavioral['labels'].isin(['face', 'house'])
conditions = conditions[condition_mask]

# Confirm that we now have 2 conditions
print(conditions.unique())

# Record these as an array of sessions, with fields
# for condition (face or house) and run
session = behavioral[condition_mask].to_records(index=False)
print(session.dtype.names)

**Comment by KamalakerDadi** [comment]
With ABIDE data you can do something like classification for instance predicting between Autism and Controls. You should use "DX_GROUP" from ```phenotypic``` key when you fetch abide data.

**Comment by meghalD** [comment]
Can someone help me understand the output of SVM using this link  http://nilearn.github.io/auto_examples/02_decoding/plot_haxby_anova_svm.html
I got similar points on the image but not sure what SVM does here and how it plots these activation points. What can be concluded from them? Any help will be appreciated as I am completely new to this field.

Thank You!!

**Comment by bthirion** [comment]
Note that you don't need any condition_mask when you work with abide, as there is no condition specified.
Regarding the SVM, I'm not sure I understand your question. The SVM is a classifier, so the best way to understand it is to refer to the corresponding doc: http://scikit-learn.org/stable/modules/svm.html
You can understand this as a black box that generate an estimate of a class label when you git it data.
Then you may want to introspect it, by looking at the associated `coefs_`.

**Comment by meghalD** [comment]
> Note that you don't need any condition_mask when you work with abide, as there is no condition specified.
> Regarding the SVM, I'm not sure I understand your question. The SVM is a classifier, so the best way to understand it is to refer to the corresponding doc: http://scikit-learn.org/stable/modules/svm.html
> You can understand this as a black box that generate an estimate of a class label when you git it data.
> Then you may want to introspect it, by looking at the associated `coefs_`.

I have dataset all downloaded in a directory. Is it possible to fetch it rather than using fetch_abide_pcp function again and again for the entire data. I tried using glob and saving all .nii.gz files in a list and tried to fit it in ICA using the following code:

for filenames_func in glob.iglob('/media/iab/disk_a/meghal/acertaabide/data/functionals/cpac/filt_global/func_preproc/*.nii.gz',recursive=True):
    func.append(filenames_func)
from nilearn import decomposition
canica = decomposition.CanICA(n_components=20, mask_strategy='background')
canica.fit(func)

#it gives errors like  
 File "/usr/local/lib/python3.5/dist-packages/sklearn/externals/joblib/parallel.py", line 131, in __call__
    return [func(*args, **kwargs) for func, args, kwargs in self.items]
  File "/usr/local/lib/python3.5/dist-packages/sklearn/externals/joblib/parallel.py", line 131, in <listcomp>
    return [func(*args, **kwargs) for func, args, kwargs in self.items]
  File "/usr/local/lib/python3.5/dist-packages/nilearn/masking.py", line 423, in compute_background_mask
    data_imgs = _utils.check_niimg(data_imgs)
  File "/usr/local/lib/python3.5/dist-packages/nilearn/_utils/niimg_conversions.py", line 271, in check_niimg
    niimg = load_niimg(niimg, dtype=dtype)
  File "/usr/local/lib/python3.5/dist-packages/nilearn/_utils/niimg.py", line 116, in load_niimg
    dtype = _get_target_dtype(niimg.get_data().dtype, dtype)
  File "/usr/local/lib/python3.5/dist-packages/nibabel/dataobj_images.py", line 202, in get_data
    data = np.asanyarray(self._dataobj)
  File "/home/iab/.local/lib/python3.5/site-packages/numpy/core/numeric.py", line 544, in asanyarray
    return array(a, dtype, copy=False, order=order, subok=True)
  File "/usr/local/lib/python3.5/dist-packages/nibabel/arrayproxy.py", line 293, in __array__
    raw_data = self.get_unscaled()
  File "/usr/local/lib/python3.5/dist-packages/nibabel/arrayproxy.py", line 288, in get_unscaled
    mmap=self._mmap)
  File "/usr/local/lib/python3.5/dist-packages/nibabel/volumeutils.py", line 524, in array_from_file
    n_read = infile.readinto(data_bytes)
  File "/usr/lib/python3.5/gzip.py", line 274, in read
    return self._buffer.read(size)
  File "/usr/lib/python3.5/_compression.py", line 68, in readinto
    data = self.read(len(byte_view))
  File "/usr/lib/python3.5/gzip.py", line 480, in read
    raise EOFError("Compressed file ended before the "
EOFError: Compressed file ended before the end-of-stream marker was reached

Can someone explain what can be the matter ? I dont want to download entire dataset again and again to just run SVM or ICA.

**Comment by meghalD** [comment]
> The compressed file you got is not complete: you obtained only part of the zip file. When trying to unzip the image reader raises an error. You should remove the existing image file and re-download it.
> HTH

I downloaded it again but got the same error. 
Can someone suggest me how to work with abide dataset specifically for Dictionary Learning. When I try to fit my data it throws and error that NiftiMasker does nto accept 'template' but 'background' or 'epi' only. I want the masking to be MNI152 template.

**Comment by meghalD** [comment]
Hello everyone ! 
Thanks for all the help earlier. I am facing another issue. It would be great if someone can help me. 

1) I have downloaded entire ABIDE preprocessed data including func_preproc, func_mask and rois_cc200. I wanted to perform dictionary learning and used this link as a reference http://nilearn.github.io/auto_examples/03_connectivity/plot_extract_regions_dictlearning_maps.html

with abide dataset I feel confounds is equivalent to rois_cc200. If not how do I calculate it or fetch it automatically.

2)Another doubt is that the data is quite huge for me to download it again and again. I have saved it in memory and want to fetch these functional files and rois_cc200 files in the format nilearn requires it to be. How do I do that. ?
